xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,725 @@
+        break;
+      }
+    }
+
+    // Field clusterName found in the in-lined RouteConfiguration, if exists.
+    String clusterName = null;
+    // RouteConfiguration name to be used as the resource name for RDS request.
+    String rdsRouteConfigName = null;
+    if (requestedHttpConnManager != null) {
+      if (requestedHttpConnManager.hasRouteConfig()) {
+        RouteConfiguration rc = requestedHttpConnManager.getRouteConfig();
+        clusterName = processRouteConfig(rc);
+        if (clusterName == null) {
+          invalidData = true;
+        }
+      } else if (requestedHttpConnManager.hasRds()) {
+        Rds rds = requestedHttpConnManager.getRds();
+        rdsRouteConfigName = rds.getRouteConfigName();
+      }
+      // Else impossible as we have already validated all HttpConnectionManager messages.
+    }
+
+    if (invalidData) {
+      adsStream.sendNackRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName),
+          ldsResponse.getNonce(),
+          "HttpConnectionManager contains invalid information for gRPC's usage");
+      return;
+    }
+    adsStream.sendAckRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName),
+        ldsResponse.getVersionInfo(), ldsResponse.getNonce());
+
+    // Remove RDS cache entries for RouteConfigurations not referenced by this LDS response.
+    // LDS responses represents the state of the world, RouteConfigurations not referenced
+    // by this LDS response are those no longer exist.
+    routeConfigNamesToClusterNames.keySet().retainAll(routeConfigs);
+
+    // Process the requested Listener if exists, either extract cluster information from in-lined
+    // RouteConfiguration message or send an RDS request for dynamic resolution.
+    if (clusterName != null) {
+      // Found clusterName in the in-lined RouteConfiguration.
+      ConfigUpdate configUpdate = ConfigUpdate.newBuilder().setClusterName(clusterName).build();
+      configWatcher.onConfigChanged(configUpdate);
+    } else if (rdsRouteConfigName != null) {
+      // Update the RDS resource we wish to request. An RDS request may not be necessary, but
+      // we need to keep what we request updated in case of notifying watcher upon receiving
+      // an RDS response for updating the requested resource.
+      adsStream.rdsResourceName = rdsRouteConfigName;
+      // First look up the RDS cache to see if we had received an RDS response containing the
+      // desired RouteConfiguration previously. Otherwise, send an RDS request for dynamic
+      // resolution.
+      if (routeConfigNamesToClusterNames.containsKey(rdsRouteConfigName)) {
+        ConfigUpdate configUpdate =
+            ConfigUpdate.newBuilder()
+                .setClusterName(routeConfigNamesToClusterNames.get(rdsRouteConfigName))
+                .build();
+        configWatcher.onConfigChanged(configUpdate);
+      } else {
+        adsStream.sendXdsRequest(ADS_TYPE_URL_RDS, ImmutableList.of(rdsRouteConfigName));
```

Comment:
On line 324 you copied `rdsRouteConfigName` into `adsStream.rdsResourceName` so you can use `adsStream.rdsResourceName` here instead. So in all cases `adsStream.sendXdsRequest` can just use the member variable `adsStream.rdsResourceName` , right?

Reply:
`sendXdsRequest(...)` method, as the spec mentions, it will set `adsStream.rdsResourceName` to memorize the most recently requested RDS resource.

Line 324 is special here (That's why I had a couple of sentences to explain it). This is the case when we received an LDS response that asks to send RDS for some resource that already exists in local cache. We can directly get the result from cache without sending an RDS request. However, we would still need to update `adsStream.rdsResourceName` as if we had sent the request. Otherwise, if we received an RDS update for this resource later, we won't know it is the one we are currently looking for.